File: 프로그래머스/3/42892. 길 찾기 게임/길 찾기 게임.py
import sys
sys.setrecursionlimit(10**6)

class node:
    def __init__(self, info):
        self.number = info[2]
        self.data = info[:2]
        self.left = None
        self.right = None
        
def addNode(root, info):
    if info[0] > root.data[0]:
        if not root.right: root.right = node(info)
        else: addNode(root.right, info)
    elif info[0] < root.data[0]:
        if not root.left: root.left = node(info)
        else: addNode(root.left, info)
    
def solution(nodeinfo):
    answer = [[], []]
    nodeidx = dict()
    
    # 노드 번호를 nodeinfo에 추가하기
    for idx, info in enumerate(nodeinfo):
        info.append(idx+1)
    
    # 노드 정렬하기
    nodeinfo.sort(key=lambda x:(-x[1], x[0]))
    
    # 이진 트리 만들기
    root = node(nodeinfo[0])
    for info in nodeinfo[1:]:
        addNode(root, info)
        
    
    # 전위 순회
    def preorder(root, order):
        if root != None:
            order.append(root.number)
            preorder(root.left, order)
            preorder(root.right, order)

    # 후위 순회
    def postorder(root, order):
        if root != None:
            postorder(root.left, order)
            postorder(root.right, order)
            order.append(root.number)
            
    
    preorder(root, answer[0])
    postorder(root, answer[1])
    return answer

File: 프로그래머스/3/42892. 길 찾기 게임/길 찾기 게임.java
import java.util.*;

class Solution {
    static class Node implements Comparable {
        int idx;
        int x, y;
        Node left, right;
        
        public Node(int idx, int x, int y, Node left, Node right) {
            this.idx = idx;
            this.x = x;
            this.y = y;
            this.left = left;
            this.right = right;
        }
        
        public int compareTo(Object obj) {
            Node n = (Node) obj;
            if (this.y == n.y) {
                return this.x - n.x;
            }
            return n.y - this.y;
        }
        
        @Override
        public String toString() {
           return String.format("Node{idx=%d, x=%d, y=%d, left=%s, right=%s}", 
                               idx, x, y, 
                               left != null ? "Node(" + left.idx + ")" : "null",
                               right != null ? "Node(" + right.idx + ")" : "null");
        }
    }
    
    static ArrayList<Integer> preOrderList;
    static ArrayList<Integer> postOrderList;
    
    public int[][] solution(int[][] nodeinfo) {        
        ArrayList<Node> nodeList = new ArrayList<>();
        for (int i = 0; i < nodeinfo.length; i++) {
            nodeList.add(new Node(i+1, nodeinfo[i][0], nodeinfo[i][1], null, null));
        }
        
        Collections.sort(nodeList);
        
        // 이진 트리 만들기
        Node root = nodeList.get(0);
        for (int i = 1; i < nodeList.size(); i++) {
            Node parent = root;
            Node curNode = nodeList.get(i);
            
            while (true) {
                if (parent.x > curNode.x) {
                    if (parent.left == null) {
                        parent.left = curNode;
                        break;
                    }
                    else parent = parent.left;
                } else {
                    if (parent.right == null) {
                        parent.right = curNode;
                        break;
                    }
                    else parent = parent.right;
                }
            }
        }
        
        preOrderList = new ArrayList<>();
        postOrderList = new ArrayList<>();
        preorder(root);
        postorder(root);
        
        int[][] answer = new int[2][nodeinfo.length];
        answer[0] = preOrderList.stream().mapToInt(Integer::intValue).toArray();
        answer[1] = postOrderList.stream().mapToInt(Integer::intValue).toArray();
        
        return answer;
    }
    
    public static void preorder(Node cur) {
        if (cur == null) return;
        
        preOrderList.add(cur.idx);
        preorder(cur.left);
        preorder(cur.right);
    }
    
    public static void postorder(Node cur) {
        if (cur == null) return;
        
        postorder(cur.left);
        postorder(cur.right);
        postOrderList.add(cur.idx);
    }
}